### PR TITLE
Add Project.OutputRefFilePath and ProjectInfo.OutputRefFilePath

### PR DESCRIPTION
**Pull request by DustinCampbell**
This change adds an OutputRefFilePath path to the workspace API in support the compiler's /refout feature. Keeping track of this path in the workspace is necessary for resolving a command line's metadata references to project references.

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/ProjectInfo.cs [hunk 1]
```diff
@@ -445,6 +498,7 @@ public void WriteTo(ObjectWriter writer)
                 writer.WriteString(Language);
                 writer.WriteString(FilePath);
                 writer.WriteString(OutputFilePath);
+                writer.WriteString(OutputRefFilePath);
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/ProjectInfo.cs]: @heejaechang Does this need a bump of some version number somewhere?

File: src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs [hunk 2]
```diff
@@ -411,6 +414,16 @@ public ProjectState UpdateOutputPath(string outputFilePath)
...
         }
 
+        public ProjectState UpdateOutputRefPath(string outputRefFilePath)
+        {
+            if (outputRefFilePath == this.OutputRefFilePath)
+            {
+                return this;
+            }
+
+            return this.With(projectInfo: this.ProjectInfo.WithOutputFilePath(outputRefFilePath).WithVersion(this.Version.GetNewerVersion()));
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs]: I assume you want this.ProjectInfo.WithOutput**Ref**Path here.

File: src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs [hunk 3]
```diff
@@ -411,6 +414,16 @@ public ProjectState UpdateOutputPath(string outputFilePath)
             return this.With(projectInfo: this.ProjectInfo.WithOutputFilePath(outputFilePath).WithVersion(this.Version.GetNewerVersion()));
         }
 
+        public ProjectState UpdateOutputRefPath(string outputRefFilePath)
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs]: "UpdateOutput**Ref**FilePath" to be consistent with everything else?

File: src/Workspaces/Core/Portable/Workspace/Solution/ProjectInfo.cs (hunk 1 above)
Comment (heejaechang) [reply src/Workspaces/Core/Portable/Workspace/Solution/ProjectInfo.cs]: no, the one who persist this data need to. right now, there is no one who persist this data.

**Comment by heejaechang** [comment]
@DustinCampbell you need to change this as well to handle the case.
http://source.roslyn.io/#Microsoft.CodeAnalysis.Remote.Workspaces/Services/SolutionCreator.cs,234

the above is for incremental update of OOP solution. what you did in this PR will only work for creating new solution from scratch but not for incremental update.

also, you can add test here to cover incremental update case.
http://source.roslyn.io/#Roslyn.VisualStudio.Next.UnitTests/Services/SolutionServiceTests.cs,155

**Comment by heejaechang** [comment]
also, I am not sure whether changing that should cause re-analyzing project or not. I can't find place where it updating compiler options. but assuming that you update version, I think you want to cause solution crawler to re-analyze projects? in that case, you need to add it in here
http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/SolutionCrawler/WorkCoordinator.cs,494

right now, it doesn't blindly just run solution crawler for any project info change since re-running solution crawler for projects are really expensive.

**Comment by heejaechang** [comment]
so we have this in Workspace.TryApplyChanges
http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Workspace.cs,1602

just realized we don't have one for ProjectInfo. not sure why. is the assumption everything in projectinfo should exist in either compilationoption or parseoption?

File: src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs (hunk 2 above)
Comment (DustinCampbell) [reply src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs]: Thanks! Missed that when I re-did this change on dev15.7.x.

File: src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs (hunk 3 above)
Comment (DustinCampbell) [reply src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs]: Thanks! Missed that when I re-did this change on dev15.7.x.

Comment (DustinCampbell) [reply src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs]: Actually, this is consistent with `UpdateOutputPath` right above this method.

**Comment by DustinCampbell** [comment]
> so we have this in Workspace.TryApplyChanges
http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Workspace.cs,1602

> just realized we don't have one for ProjectInfo. not sure why. is the assumption everything in projectinfo should exist in either compilationoption or parseoption?

No. ProjectInfo exposes CompilationOptions and ParseOptions. All other properties exposed on ProjectInfo are generally the properties that aren't included in CompilationOptions and ParseOptions (such as OutputFIlePath).

Adding `ApplyProjectInfoChanged` is kind of outside of the scope of this PR.

**Comment by DustinCampbell** [comment]
@heejaechang:

> you need to change this as well to handle the case.
http://source.roslyn.io/#Microsoft.CodeAnalysis.Remote.Workspaces/Services/SolutionCreator.cs,234

Thanks!

File: src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs (hunk 3 above)
Comment (DustinCampbell) [reply src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs]: FWIW, I'll go ahead and change this particular method.

**Comment by DustinCampbell** [comment]
@heejaechang:

> also, I am not sure whether changing that should cause re-analyzing project or not. I can't find place where it updating compiler options. but assuming that you update version, I think you want to cause solution crawler to re-analyze projects? in that case, you need to add it in here
http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/SolutionCrawler/WorkCoordinator.cs,494

> right now, it doesn't blindly just run solution crawler for any project info change since re-running solution crawler for projects are really expensive.

Since the WorkCoordinator doesn't seem to consider changes to the `OutputFilePath`, I don't expect it needs to consider changes to the `OutputRefFilePath`. (Note that `OutputFilePath` is not part of compilation options.)

File: src/Workspaces/Core/Portable/Workspace/Solution/ProjectInfo.cs (hunk 1 above)
Comment (DustinCampbell) [reply src/Workspaces/Core/Portable/Workspace/Solution/ProjectInfo.cs]: excellent -- thanks for checking @heejaechang!

**Pull request merged**